File: CodingSamples/Java/Platform/AsyncParallelTest/Program2.java
import java.util.stream.*;
import java.util.concurrent.*;

class Program {
	
	static class Computation {

		private long start;

		public long compute(int begin, int end) {
			start = System.currentTimeMillis();
			return IntStream.range(begin, end)
						.mapToLong(Worker::doWork)
						.sum();
		}

		public CompletableFuture<Long> computeAsync(int begin, int end) {
			//the supplied expression is evaluated by a pooled thread 
			//allowing the caller thread to resume its own execution
			//and to accept the result of evaluation at a time in
			//future when the evaluation is completed
			return CompletableFuture.supplyAsync(() -> compute(begin, end));
		}

		public double time() {
			long stop = System.currentTimeMillis();
			return (stop - start) / 1000.0;
		}

	}

	public static void main(String[] args) throws Exception {
		int n = Integer.parseInt(args[0]);
		System.out.print("Computing...");
		var c = new Computation();
		var job = c.computeAsync(1, n + 1)
					.thenAccept(r -> {
						System.out.println("Done!");
						System.out.printf("Result = %d, computed in %.3f seconds.%n", r, c.time());
					});
		while(!job.isDone()){
			System.out.print(".");
			Thread.sleep(500);
		}
	}
}

File: CodingSamples/DotNet/Platform/AsyncParallelTest/DemoApp/Program2.cs
class Computation
{
    private System.Diagnostics.Stopwatch clock = new();

    public long Compute(int begin, int end)
    {
        clock.Start();
        return Enumerable.Range(begin, end - begin)
                .Select(Worker.DoWork)
                .Sum();
    }

    public Task<long> ComputeAsync(int begin, int end)
    {
        //a method passed to Run is invoked by a pooled thread
        //allowing the caller thread to resume execution and 
        //obtain the result of this invocation when it is completed
        return Task<long>.Run(() => Compute(begin, end));
    }

    public double Time()
    {
        clock.Stop();
        return clock.Elapsed.TotalSeconds;
    }
}

class Program
{

    static Task DoComputation(int limit)
    {
        Console.Write("Computing...");
        var c = new Computation();
        return c.ComputeAsync(1, limit + 1)
                .ContinueWith(previous => 
                {
                    long r = previous.Result;
                    Console.WriteLine("Done.");
                    Console.WriteLine("Result = {0}, computed in {1:0.000} seconds.", r, c.Time());                    
                });

    }

    static void Main(string[] args)
    {
        int n = int.Parse(args[0]);
        Task job = DoComputation(n);
        while(!job.IsCompleted)
        {
            Console.Write(".");
            Thread.Sleep(500);
        }
    }
}
